File: studio/AuraStudio/Sources/AuraStudio/Services/FanartTVClient.swift
import Foundation

/// Cliente de fanart.tv (D-203): caratulas y arte de disco en alta
/// resolucion, fuente OPCIONAL con API key propia (a diferencia de
/// MusicBrainz/Cover Art Archive/LRCLIB). La key se lee del Keychain
/// via `APIKeyStore` -- nunca se guarda ni se pasa por UserDefaults.
///
/// Endpoint confirmado contra la API real (agosto 2026):
/// `GET /v3/music/albums/{release-group-id}?api_key=...` -- indexa por
/// RELEASE GROUP (el album), no por release (la edicion especifica).
/// Sin key configurada no tiene sentido intentar la llamada (fanart.tv
/// siempre devuelve 401): se devuelve nil de una, mismo criterio de
/// "mejor esfuerzo" que ya usan los demas clientes.
struct FanartTVClient {
    private struct AlbumsResponse: Decodable {
        let albums: [String: AlbumImages]?
    }

    private struct AlbumImages: Decodable {
        let albumcover: [Image]?
    }

    private struct Image: Decodable {
        let url: String
    }

    private struct ArtistResponse: Decodable {
        let artistthumb: [Image]?
        let artistbackground: [Image]?
    }

    private struct MovieResponse: Decodable {
        let movieposter: [Image]?
    }

    private struct TVResponse: Decodable {
        let tvposter: [Image]?
    }

    private let session: URLSession
    private let baseURL: URL
    /// Raiz `/v3` (ST-032): `baseURL` sigue apuntando a `/music/albums`
    /// por compatibilidad con D-203; artistas, peliculas y series
    /// cuelgan de esta.
    private let rootURL: URL

    /// De donde sale la key: el Keychain (`APIKeyStore`) en la app;
    /// inyectable en tests para no tocar el llavero real.
    private let apiKeyProvider: @Sendable () -> String?

    init(session: URLSession = .shared,
         baseURL: URL = URL(string: "https://webservice.fanart.tv/v3/music/albums")!,
         rootURL: URL = URL(string: "https://webservice.fanart.tv/v3")!,
         apiKeyProvider: @escaping @Sendable () -> String? = { APIKeyStore.load(for: .fanartTV) }) {
        self.session = session
        self.baseURL = baseURL
        self.rootURL = rootURL
        self.apiKeyProvider = apiKeyProvider
    }

    /// GET autenticado; 404 = "fanart.tv no lo tiene" (nil, no error);
    /// sin key = nil sin tocar la red.
    private func fetchJSON<T: Decodable>(_ type: T.Type, path: String) async throws -> T? {
        guard let apiKey = apiKeyProvider() else { return nil }
        var components = URLComponents(url: rootURL.appendingPathComponent(path), resolvingAgainstBaseURL: false)!
        components.queryItems = [URLQueryItem(name: "api_key", value: apiKey)]
        var request = URLRequest(url: components.url!)
        request.setValue(MusicBrainzClient.userAgent, forHTTPHeaderField: "User-Agent")
        let (data, response) = try await session.data(for: request)
        if let http = response as? HTTPURLResponse, http.statusCode == 404 { return nil }
        try MusicBrainzClient.validate(response)
        return try JSONDecoder().decode(T.self, from: data)
    }

    private func download(_ urlString: String?) async throws -> Data? {
        guard let urlString, let url = URL(string: urlString) else { return nil }
        let (data, response) = try await session.data(from: url)
        try MusicBrainzClient.validate(response)
        return data
    }

    /// ST-032: foto de artista (`artistthumb`, cuadrada, ~1000 px) por
    /// MusicBrainz artist ID.
    func fetchArtistThumb(musicBrainzArtistID: String) async throws -> Data? {
        let decoded = try await fetchJSON(ArtistResponse.self, path: "music/\(musicBrainzArtistID)")
        return try await download(decoded?.artistthumb?.first?.url)
    }

    /// ST-033: poster de pelicula. fanart.tv acepta el ID de TMDB o el de
    /// IMDb (`tt...`) en la misma ruta.
    func fetchMoviePoster(tmdbOrIMDbID: String) async throws -> Data? {
        let decoded = try await fetchJSON(MovieResponse.self, path: "movies/\(tmdbOrIMDbID)")
        return try await download(decoded?.movieposter?.first?.url)
    }

    /// ST-033: poster de serie, por ID de TheTVDB (el unico que fanart.tv
    /// indexa para TV -- se obtiene via TMDB `external_ids`).
    func fetchTVPoster(tvdbID: String) async throws -> Data? {
        let decoded = try await fetchJSON(TVResponse.self, path: "tv/\(tvdbID)")
        return try await download(decoded?.tvposter?.first?.url)
    }

    func fetchAlbumCover(releaseGroupID: String) async throws -> Data? {
        guard let apiKey = apiKeyProvider() else { return nil }

        var components = URLComponents(url: baseURL.appendingPathComponent(releaseGroupID), resolvingAgainstBaseURL: false)!
        components.queryItems = [URLQueryItem(name: "api_key", value: apiKey)]

        var request = URLRequest(url: components.url!)
        request.setValue(MusicBrainzClient.userAgent, forHTTPHeaderField: "User-Agent")

        let (data, response) = try await session.data(for: request)
        // fanart.tv no tiene el album: 404 normal, no un error a
        // reportar (la mayoria de los albumes simplemente no estan en
        // su base).
        if let http = response as? HTTPURLResponse, http.statusCode == 404 { return nil }
        try MusicBrainzClient.validate(response)

        let decoded = try JSONDecoder().decode(AlbumsResponse.self, from: data)
        guard let urlString = decoded.albums?[releaseGroupID]?.albumcover?.first?.url,
              let imageURL = URL(string: urlString) else { return nil }

        let (imageData, imageResponse) = try await session.data(from: imageURL)
        try MusicBrainzClient.validate(imageResponse)
        return imageData
    }
}

File: studio/windows/AuraStudio.Core/Networking/FanartTVClient.cs
using System.Net;
using System.Net.Http;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace AuraStudio.Core.Networking;

/// <summary>
/// Cliente de fanart.tv (D-203): carátulas y arte de disco en alta
/// resolución, fuente OPCIONAL con API key propia. Sin key configurada
/// no tiene sentido intentar la llamada (fanart.tv siempre devuelve 401):
/// se devuelve null de una, mismo criterio de "mejor esfuerzo" que ya usan
/// los demás clientes.
/// </summary>
public sealed class FanartTVClient
{
    private sealed class AlbumsResponse
    {
        [JsonPropertyName("albums")] public Dictionary<string, AlbumImages>? Albums { get; init; }
    }

    private sealed class AlbumImages
    {
        [JsonPropertyName("albumcover")] public List<Image>? AlbumCover { get; init; }
    }

    private sealed class Image
    {
        [JsonPropertyName("url")] public string Url { get; init; } = "";
    }

    private sealed class ArtistResponse
    {
        [JsonPropertyName("artistthumb")] public List<Image>? ArtistThumb { get; init; }
        [JsonPropertyName("artistbackground")] public List<Image>? ArtistBackground { get; init; }
    }

    private sealed class MovieResponse
    {
        [JsonPropertyName("movieposter")] public List<Image>? MoviePoster { get; init; }
    }

    private sealed class TVResponse
    {
        [JsonPropertyName("tvposter")] public List<Image>? TvPoster { get; init; }
    }

    private readonly HttpClient _http;
    private readonly string _rootURL;
    private readonly IApiKeyStore _apiKeyStore;

    public FanartTVClient(
        HttpClient? http = null,
        IApiKeyStore? apiKeyStore = null,
        string rootURL = "https://webservice.fanart.tv/v3")
    {
        _http = http ?? new HttpClient();
        _rootURL = rootURL;
        _apiKeyStore = apiKeyStore ?? new SimpleApiKeyStore();
    }

    /// <summary>GET autenticado; 404 = "fanart.tv no lo tiene" (null, no error).</summary>
    private async Task<T?> FetchJSONAsync<T>(string path, CancellationToken ct = default) where T : class
    {
        var apiKey = _apiKeyStore.Load("fanarttv");
        if (apiKey == null) return null;

        var url = $"{_rootURL}/{path}?api_key={Uri.EscapeDataString(apiKey)}";
        using var request = new HttpRequestMessage(HttpMethod.Get, url);
        request.Headers.TryAddWithoutValidation("User-Agent", HttpUserAgent.Value);

        using var response = await _http.SendAsync(request, ct).ConfigureAwait(false);
        if (response.StatusCode == HttpStatusCode.NotFound) return null;
        HttpResponseGuard.EnsureSuccess(response);

        var json = await response.Content.ReadAsStringAsync(ct).ConfigureAwait(false);
        return JsonSerializer.Deserialize<T>(json, new JsonSerializerOptions { PropertyNameCaseInsensitive = true });
    }

    /// <summary>Descarga la imagen desde la URL dada.</summary>
    private async Task<byte[]?> DownloadAsync(string? urlString, CancellationToken ct = default)
    {
        if (string.IsNullOrEmpty(urlString)) return null;
        using var request = new HttpRequestMessage(HttpMethod.Get, urlString);
        using var response = await _http.SendAsync(request, ct).ConfigureAwait(false);
        if (!response.IsSuccessStatusCode) return null;
        return await response.Content.ReadAsByteArrayAsync(ct).ConfigureAwait(false);
    }

    /// <summary>ST-032: foto de artista (artistthumb, cuadrada, ~1000 px) por MusicBrainz artist ID.</summary>
    public async Task<byte[]?> FetchArtistThumbAsync(string musicBrainzArtistID, CancellationToken ct = default)
    {
        var decoded = await FetchJSONAsync<ArtistResponse>($"music/{musicBrainzArtistID}", ct).ConfigureAwait(false);
        return await DownloadAsync(decoded?.ArtistThumb?.FirstOrDefault()?.Url, ct).ConfigureAwait(false);
    }

    /// <summary>ST-033: póster de película. fanart.tv acepta TMDB ID o IMDb ID en la misma ruta.</summary>
    public async Task<byte[]?> FetchMoviePosterAsync(string tmdbOrIMDbID, CancellationToken ct = default)
    {
        var decoded = await FetchJSONAsync<MovieResponse>($"movies/{tmdbOrIMDbID}", ct).ConfigureAwait(false);
        return await DownloadAsync(decoded?.MoviePoster?.FirstOrDefault()?.Url, ct).ConfigureAwait(false);
    }

    /// <summary>ST-033: póster de serie, por ID de TheTVDB.</summary>
    public async Task<byte[]?> FetchTVPosterAsync(string tvdbID, CancellationToken ct = default)
    {
        var decoded = await FetchJSONAsync<TVResponse>($"tv/{tvdbID}", ct).ConfigureAwait(false);
        return await DownloadAsync(decoded?.TvPoster?.FirstOrDefault()?.Url, ct).ConfigureAwait(false);
    }

    /// <summary>Cover de álbum por MusicBrainz release-group ID.</summary>
    public async Task<byte[]?> FetchAlbumCoverAsync(string releaseGroupID, CancellationToken ct = default)
    {
        var apiKey = _apiKeyStore.Load("fanarttv");
        if (apiKey == null) return null;

        var url = $"https://webservice.fanart.tv/v3/music/albums/{releaseGroupID}?api_key={Uri.EscapeDataString(apiKey)}";
        using var request = new HttpRequestMessage(HttpMethod.Get, url);
        request.Headers.TryAddWithoutValidation("User-Agent", HttpUserAgent.Value);

        using var response = await _http.SendAsync(request, ct).ConfigureAwait(false);
        if (response.StatusCode == HttpStatusCode.NotFound) return null;
        HttpResponseGuard.EnsureSuccess(response);

        var json = await response.Content.ReadAsStringAsync(ct).ConfigureAwait(false);
        var decoded = JsonSerializer.Deserialize<AlbumsResponse>(json, new JsonSerializerOptions { PropertyNameCaseInsensitive = true });
        var urlString = decoded?.Albums?.GetValueOrDefault(releaseGroupID)?.AlbumCover?.FirstOrDefault()?.Url;
        return await DownloadAsync(urlString, ct).ConfigureAwait(false);
    }

    private sealed class SimpleApiKeyStore : IApiKeyStore
    {
        public string? Load(string service) => null;
    }
}